lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/json/JsonDeserializer.java

```diff
@@ -604,48 +633,58 @@ void decodeValue(JsonParser parser, BlockBuilder builder)
-            }
-            else if (type instanceof VarcharType varcharType) {
-                type.writeSlice(builder, truncateToLength(Slices.utf8Slice(value), varcharType));
-            }
-            else if (type instanceof CharType charType) {
-                type.writeSlice(builder, truncateToLengthAndTrimSpaces(Slices.utf8Slice(value), charType));
+            try {
+                if (BOOLEAN.equals(type)) {
+                    type.writeBoolean(builder, Boolean.parseBoolean(value));
+                }
+                else if (BIGINT.equals(type)) {
+                    type.writeLong(builder, Long.parseLong(value));
+                }
+                else if (INTEGER.equals(type)) {
+                    INTEGER.writeLong(builder, Integer.parseInt(value));
+                }
+                else if (SMALLINT.equals(type)) {
+                    SMALLINT.writeLong(builder, Short.parseShort(value));
+                }
+                else if (TINYINT.equals(type)) {
+                    TINYINT.writeLong(builder, Byte.parseByte(value));
+                }
+                else if (type instanceof DecimalType decimalType) {
+                    writeDecimal(decimalType, builder, value);
+                }
+                else if (REAL.equals(type)) {
+                    REAL.writeLong(builder, floatToRawIntBits(Float.parseFloat(value)));
+                }
+                else if (DOUBLE.equals(type)) {
+                    writeDouble(builder, Double.parseDouble(value));
+                }
+                else if (DATE.equals(type)) {
+                    DATE.writeLong(builder, parseHiveDate(value).toEpochDay());
+                }
+                else if (type instanceof TimestampType timestampType) {
+                    DecodedTimestamp timestamp = timestampParser.apply(value);
+                    createTimestampEncoder(timestampType, UTC).write(timestamp, builder);
+                }
+                else if (VARBINARY.equals(type)) {
+                    writeSlice(VARBINARY, builder, VarbinaryDecoder.parseBinary(value, charsetDecoder));
+                }
+                else if (type instanceof VarcharType varcharType) {
+                    writeSlice(varcharType, builder, truncateToLength(Slices.utf8Slice(value), varcharType));
+                }
+                else if (type instanceof CharType charType) {
+                    writeSlice(charType, builder, truncateToLengthAndTrimSpaces(Slices.utf8Slice(value), charType));
+                }
+                else {
+                    throw new UnsupportedOperationException("Unsupported map key type: " + type);
+                }
             }
-            else {
-                throw new UnsupportedOperationException("Unsupported map key type: " + type);
+            catch (TrinoException e) {
+                if (e.getErrorCode() == GENERIC_INTERNAL_ERROR.toErrorCode()) {
```

Comment:
This is a bug somewhere else. Is this covered by tests?

We should fix the bug instead of adding this code here.

Reply:
This error will be thrown in a couple places either directly by `RealType#writeLong` if the value is not a valid float or by `AbstractIntType#checkValidValue` when checking if the value is a valid int for int, smallint, and tinyint. So this should be considered a parse/bad data error. Alternatively, we could just throw a `BAD_DATA` error at the sources.